src/decode/lzma2.rs

```diff
@@ -111,85 +78,146 @@ where
+            "LZMA2 uncompressed block {{ unpacked_size: {}, reset_dict: {} }}",
+            unpacked_size,
+            reset_dict
+        );
+
+        if reset_dict {
+            accum.reset()?;
+        }
 
-            lzma_info!("Properties {{ lc: {}, lp: {}, pb: {} }}", lc, lp, pb);
-            LzmaProperties { lc, lp, pb }
-        } else {
-            decoder.lzma_props
-        };
+        let mut buf = vec![0; unpacked_size];
+        input.read_exact(buf.as_mut_slice()).map_err(|e| {
+            error::Error::LzmaError(format!(
+                "LZMA2 expected {} uncompressed bytes: {}",
+                unpacked_size, e
+            ))
+        })?;
+        accum.append_bytes(buf.as_slice());
 
-        decoder.reset_state(new_props);
+        Ok(())
     }
 
-    decoder.set_unpacked_size(Some(unpacked_size + accum.len() as u64));
-
-    let mut taken = input.take(packed_size);
-    let mut rangecoder = rangecoder::RangeDecoder::new(&mut taken)
-        .map_err(|e| error::Error::LzmaError(format!("LZMA input too short: {}", e)))?;
-    decoder.process(accum, &mut rangecoder)
-}
-
-fn parse_uncompressed<R, W>(
-    accum: &mut lzbuffer::LzAccumBuffer<W>,
-    input: &mut R,
-    reset_dict: bool,
-) -> error::Result<()>
-where
-    R: io::BufRead,
-    W: io::Write,
-{
-    let unpacked_size = input
-        .read_u16::<BigEndian>()
-        .map_err(|e| error::Error::LzmaError(format!("LZMA2 expected unpacked size: {}", e)))?;
-    let unpacked_size = (unpacked_size as usize) + 1;
-
-    lzma_info!(
-        "LZMA2 uncompressed block {{ unpacked_size: {}, reset_dict: {} }}",
-        unpacked_size,
-        reset_dict
-    );
+    /// Creates a new object ready for decompressing data that it's given.
+    pub fn new() -> Lzma2Decoder {
+        Lzma2Decoder {
+            lzma_state: DecoderState::new(
+                LzmaProperties {
+                    lc: 0,
+                    lp: 0,
+                    pb: 0,
+                },
+                None,
+            ),
+        }
+    }
 
-    if reset_dict {
-        accum.reset()?;
+    /// Performs the equivalent of replacing this decompression state with a freshly allocated copy.
+    ///
+    /// This function may not allocate memory and will attempt to reuse any previously allocated resources.
+    #[cfg(feature = "raw_decoder")]
+    pub fn reset(&mut self) {
+        self.lzma_state.reset_state(LzmaProperties {
+            lc: 0,
+            lp: 0,
+            pb: 0,
+        });
     }
 
-    let mut buf = vec![0; unpacked_size];
-    input.read_exact(buf.as_mut_slice()).map_err(|e| {
-        error::Error::LzmaError(format!(
-            "LZMA2 expected {} uncompressed bytes: {}",
-            unpacked_size, e
-        ))
-    })?;
-    accum.append_bytes(buf.as_slice());
+    /// Decompresses the input data into the output, consuming only as much input as needed and writing as much output as possible.
+    pub fn decompress<W: io::Write, R: io::BufRead>(
```

Comment:
And as this one is public I'd put it at the top as well. This may also make the diff easier for GitHub.